src/QsCompiler/CompilerOptimizations/Evaluation.fs

```diff
@@ -13,170 +13,181 @@ open Microsoft.Quantum.QsCompiler.SyntaxTree
+/// The second element is a counter that stores the remaining number of statements we evaluate.
+type private EvalState = Map<string, TypedExpression> * int
+
 /// Represents any interrupt to the normal control flow of a function evaluation.
 /// Includes return statements, errors, and (if they were added) break/continue statements.
 type private FunctionInterrupt =
 /// Represents the function invoking a return statement
 | Returned of TypedExpression
 /// Represents the function invoking a fail statement
 | Failed of TypedExpression
+/// We reached the limit of statement executions
+| OutOfStatements
 /// We were unable to evaluate the function for some reason
 | CouldNotEvaluate of string
 
-/// The current state of a function evaluation.
-/// The Ok case means that we are evaluating the function, inside normal control flow.
-/// The Ok case stores a map from all local variables to their current values.
-/// The Error case means that we are outside normal control flow, in some kind of interrupt.
-/// The Error case stores a FunctionInterrupt that describes the cause of the interrupt.
-/// I use the Result type to match the common design pattern of railway-oreiented programming.
-type private FunctionState = Result<Constants<TypedExpression>, FunctionInterrupt>
+/// A shorthand for the specific Imperative type used by several functions in this file
+type private Imp<'t> = Imperative<EvalState, 't, FunctionInterrupt>
+
+/// Represents a computation that decreases the remaining statements counter by 1.
+/// Yields an OutOfStatements interrupt if this decreases the remaining statements below 0.
+let private incrementState: Imp<Unit> = imperative {
+    let! vars, counter = getState
+    if counter < 1 then yield OutOfStatements
+    do! putState (vars, counter - 1)
+}
+
+/// Represents a computation that sets the given variables to the given values
+let private setVars callables entry: Imp<Unit> = imperative {
+    let! vars, counter = getState
+    do! putState (defineVarTuple (isLiteral callables) vars entry, counter)
+}
 
 
 /// Evaluates functions by stepping through their code
-type internal FunctionEvaluator(callables: Callables, maxRecursiveDepth: int) =
+type internal FunctionEvaluator(callables: Callables) =
 
-    /// Transforms a BoolLiteral into the corresponding bool
-    let castToBool f x =
+    /// Casts a BoolLiteral to the corresponding bool
+    let castToBool x: Imp<bool> = imperative {
         match x.Expression with
-        | BoolLiteral b -> Ok (f b)
-        | _ -> "Not a BoolLiteral: " + (printExpr x.Expression) |> CouldNotEvaluate |> Error
-
-    /// The callback for the subroutine that evaluates and simplifies an expression
-    member internal this.evaluateExpression constants expr =
-        ExpressionEvaluator(callables, constants, maxRecursiveDepth - 1).Transform expr
+        | BoolLiteral b -> return b
+        | _ -> yield CouldNotEvaluate ("Not a BoolLiteral: " + (printExpr x.Expression))
```

Comment:
Is this a case that should be reached? If so, how about renaming the function to tryAsBoolLiteral or tryEvaluateBool or something?

Reply:
I suppose the second case shouldn't be reached anymore, as the modifications to evaluateExpression mean that only literal values will be passed to castToBool. Having it yield a CouldNotEvaluate interrupt would be safer than raising an exception in case it's passed an incorrect value for some reason. Should we have the code be cautious or should it raise an exception on input that should be impossible?

Reply:
The risk with "cautious" code is that is can get a lot harder to maintain and debug, because over time, a lot of the assumptions that were originally made are no longer applicable. Writing permissive like that means that these chances can go undetected for a long time. There is definitively a benefit to writing permissive code if you have a lot of external dependencies or the code is user facing. However, given that this is an internal function in a module that is maintained by us and has no external dependencies whatsoever, I'd rather we know about and fix "minor" bugs instead of ignoring them silently.

Reply:
Fixed in latest commit